debian-bo-master/bo/source/net/scotty-2.0.2/doc/scotty.1
.TH scotty 1 "12 June 1995" "Tkined" "Tkined Reference Manual"

.SH NAME
scotty \- a Tcl interpreter with TCP/IP extensions

.SH SYNOPSIS
.B scotty
[
.B -c
.I query
]
[
.B -f
.I file
]
[
.B -n
]
[
.B -i
]

.SH DESCRIPTION

scotty is a Tcl interpreter with extensions to obtain status and
configuration information about TCP/IP networks.  After startup,
scotty evaluates the commands stored in $HOME/.scottyrc.

.SH OPTIONS

.TP
.BI "-c " command
Process 
.I command
and exit. This is useful for writing shell scripts.

.TP
.BI "-f " filename
Evaluate the commands contained in 
.I filename. 

.TP
.B -n
Don't read the file $HOME/.scottyrc.

.TP
.B -i
Initialize the interpreter to run a script for the tkined(1) editor.
This option will set up the correct search path. 

.SH VARIABLES
Scotty sets the following Tcl variables:
.TP 18
\fBargc\fR
Contains a count of the number of \fIarg\fR arguments (0 if none),
not including the name of the script file.
.TP 18
\fBargv\fR
Contains a Tcl list whose elements are the \fIarg\fR arguments,
in order, or an empty string if there are no \fIarg\fR arguments.
.TP 18
\fBargv0\fR
Contains \fIfileName\fR if it was specified.
Otherwise, contains the name by which scotty was invoked.
.TP 18
\fBtcl_interactive\fR
Contains 1 if scotty is running interactively (no
\fIfileName\fR was specified and standard input is a terminal-like
device), 0 otherwise.
.TP 18
\fBscotty_library\fR
The path to the scotty library directory. This variable is
initialized to the compiled in default. You can overwrite this
default by setting the SCOTTY_LIBRARY environment variable.
.TP 18
\fBscotty_version\fR
The scotty version number.
.TP 18
\fBscotty_arch\fR
A string that identifies the architecture and operating system.
Note, this will only work with Tcl version 7.5 and higher.
.TP 18
\fBscotty_RcFileName\fR
The file name that contains scotty specific startup commands.


.SH PROMPTS
When scotty is invoked interactively it normally prompts for each
command with ``\fB% \fR''.  You can change the prompt by setting the
variables \fBtcl_prompt1\fR and \fBtcl_prompt2\fR.  If variable
\fBtcl_prompt1\fR exists then it must consist of a Tcl script
to output a prompt;  instead of outputting a prompt scotty
will evaluate the script in \fBtcl_prompt1\fR.
The variable \fBtcl_prompt2\fR is used in a similar way when
a newline is typed but the current command isn't yet complete;
if \fBtcl_prompt2\fR isn't set then no prompt is output for
incomplete commands.

.SH TCL COMMANDS
scotty is based on Tcl version 7.4 and hence it understands all of the
commands provided by Tcl 7.4. The following is a description of the 
extensions provided by scotty.

.SS BASIC EVENT HANDLING
.TP
.B after \fIargs\fR
.ns
.TP
.B update \fIargs\fR
.ns
.TP
.B tkwait \fIargs\fR
.ns
.TP
.B fileevent \fIargs\fR
These commands allow basic event handling inside scotty. See the Tk
documentation for a description of how these commands work.

.SS TIME AND DATE
.TP
.B getclock \fR[\fIdate\fR]
Return the number of seconds since 1970. This is the same as the
getclock command of extended Tcl. The optional argument forces
getclock to convert \fIdate\fR. \fIdate\fR should have the format
as returned by getdate (see below).

.TP
.B getdate \fR[\fIclock\fR]
Returns the date in human readable format. The optional argument
\fIclock\fR specifies the time in seconds since 1970 which will be
converted to a format like 'Tue Apr 26 20:32:23 1994'. The command
returns the current date if no argument is given.

.SS JOB SCHEDULING
The job scheduler provides a convenient mechanism to implement Tcl
procedures that are invoked at regular intervals. The job command
allows to associate attributes to individual jobs. This can be used
to keep state information without using global variables.

.TP
.B job create \fIcommand\fR \fIinterval\fR [\fIiterations\fR]
The job create command creates a new job for the given Tcl
\fIcommand\fR. The command will be evaluated every \fIinterval\fR
milliseconds. The optional argument defines the number of
\fIiterations\fR until the job is destroyed. A job will not
automatically destroyed if the \fIiterations\fR argument is missing.,
The job create command returns a jobs handle which can be used later
to modify the job object.
.TP
.B job current
Get the name of the currently running job. An empty string will 
be returned if no job is running. This command is useful if a job 
should modify itself.
.TP
.B job info
Returns a list of all job names. The list will be empty if there are
no jobs in this interpreter.
.TP
.B job schedule
Invoke the scheduler. This is usually done automatically. This command
may be used during long computations since the scheduler is not
preemptive.
.TP
.B job wait
The job wait command waits until all jobs have been finished. Note,
the command will wait infinite if there is at least one job that has
no iterations count set. Scotty processes events while waiting for
the jobs to complete.

.TP
.B job# status [\fIvalue\fR]
 Get or set the status of a job. Allowed status values are
\fIrunning\fR, \fIwaiting\fR, \fIsuspended\fR and
\fIexpired\fR. Setting the job state \fIsuspended\fR will take the job out
of the scheduler and setting the job state to \fIexpired\fR will
terminate the job at the next scheduling point.
.TP
.B job# command [\fIvalue\fR]
Get or set the Tcl command associated with a job.
.TP
.B job# interval [\fIvalue\fR]
Get or set the interval of a job. The value is the time interval between
two invocations and given in milliseconds.
.TP
.B job# iterations [\fIvalue\fR]
Get or set the remaining number of iterations for this job.
.TP
.B job# time
This command returns the time left til the next invocation of the job.
.TP
.B job# wait
Wait until the job is destroyed. This can happen because the number of
iterations reaches zero or because the job is terminated by setting the
job state to \fIexpired\fR.

.TP
.B job# attribute \fR[\fIname [\fIvalue\fR]\fR]
The attribute option allows to save job related data as job attributes.
If called without any arguments, all attribute names will be returned.
If called with a name argument, the current value of the attribute
will be returned. Note, non existing attributes do not produce an
error. Instead, an empty string is returned. Calling the option with
a name and a value causes value to be saved in the named attribute.

.SS EVENT LOGGING
.TP
.BI syslog " level message"
Write \fImessage\fR to the system log using priority \fIlevel\fR. You
can use the following levels:
.RS
.TP 14
.SB emergency
A panic condition.  This is normally broadcast to all users.
.TP
.SB alert
A condition that should be corrected immediately,
such as a corrupted system database.
.TP
.SB critical
Critical conditions, such as hard device errors.
.TP
.SB error
Errors.
.TP
.SB warning
Warning messages.
.TP
.SB notice
Conditions that are not error conditions,
but that may require special handling.
.TP
.SB info
Informational messages.
.TP
.SB debug
Messages that contain information
normally of use only when debugging a program.
.RE

.SS NETWORK DATABASES
The netdb command provides access to network information that is
either stored in configuration files (usually in /etc), the Network
Information Service (NIS) or in the domain name system (DNS). The
result of these commands usually reflects local configuration. If
you want to make sure to use the DNS, please use the dns command
described below.

.TP 
\fBnetdb hosts\fR
.ns
.TP 
\fBnetdb hosts name \fIaddress\fR
.ns
.TP
\fBnetdb hosts address \fIname\fR
The first command returns a list of known hosts. Each element contains
a host name and its IP address. The result is an empty list on many
systems that are configured to use the DNS to lookup hostnames.  The
second command converts an IP \fIaddress\fR into a host name and the
third command takes a host \fIname\fR and returns the corresponding IP
address.

.TP 
\fBnetdb networks\fR
.ns
.TP 
\fBnetdb networks name \fIaddress\fR
.ns
.TP
\fBnetdb networks address \fIname\fR
The first command returns a list of known networks. Each element contains
the network name and its IP network address. The seconds command returns
the name that belongs to \fIaddress\fR and the third command returns
the network address that belongs to a given \fIname\fR.

.TP 
\fBnetdb protocols\fR
.ns
.TP 
\fBnetdb protocols name \fInumber\fR
.ns
.TP
\fBnetdb protocols number \fIname\fR
The first command returns a list of known Internet protocols. Each element
contains the protocol name and its protocol number. The second command
returns the protocol name that belongs to \fInumber\fR and the third command
returns the protocol number that belongs to \fIname\fR.

.TP 
\fBnetdb services\fR
.ns
.TP 
\fBnetdb services name \fInumber\fR
.ns
.TP
\fBnetdb services number \fIname\fR
The first command returns the list of known services. The list contains
elements where each element contains the service name, the service
number and the transport protocol. The second command returns the
service name for the service given by \fInumber\fR, which must be a
list containing the service number and the transport protocol.  For
example, netdb services name "21 tcp" will return ftp.  The third
command accepts a list containing a service name and its transport
protocol and returns the protocol number. To follow the example, 
netdb services number "ftp tcp" will return 21.

.TP 
\fBnetdb sunrpcs\fR
.ns
.TP 
\fBnetdb sunrpcs name \fInumber\fR
.ns
.TP
\fBnetdb sunrpcs number \fIname\fR
The first command lists all known Sun RPCs. For each RPC, a list
containing the RPC name and its registration number is returned. The
second command returns the RPC name that belongs to \fInumber\fR and
the third command returns the RPC number that belongs to the given
\fIname\fR.

.SS INTERNET CONTROL MESSAGES (ICMP)
You need the ntping(1) program to use the icmp command since sending 
icmp packets requires root permissions.  ntping must be installed set
uid root.

.TP
\fBicmp\fR [\fIoptions\fR] \fBecho\fR \fIhosts\fR
This icmp command can be used to test the reachability of IP devices
by sending an ICMP echo request to the hosts given in \fIhosts\fR.
The command returns a list of host / round trip time pairs.

.TP
\fBicmp\fR [\fIoptions\fR] \fBmask\fR \fIhosts\fR
This icmp query returns a list of host / netmask pairs. It can
be used to locate subnetworks. The mask will be set to 0.0.0.0
if the host does not respond to mask requests.

.TP
\fBicmp\fR [\fIoptions\fR] \fBtimestamp\fR \fIhosts\fR
Ask hosts for their timestamps. This command returns an estimate of the
time difference between local time and the time on the hosts given in
\fIhosts\fR. The command returns a list of host / time pairs.

.TP
\fBicmp\fR [\fIoptions\fR] \fBttl\fR \fInum\fR \fIhosts\fR
The ttl option, which may not be supported on every unix system, 
causes scotty to send an udp packet to an unused port number with
the time to live set to \fInum\fR. The result is a list of host / 
round trip time pairs, where the host is the IP device being 
\fInum\fR hops away.

.TP
\fBicmp\fR [\fIoptions\fR] \fBtrace\fR \fInum\fR \fIhosts\fR
The trace option works similar to the ttl option but it always
returns the destination address even when the responding 
host chooses another interface with a different ip address to send 
the reply. This can be used to trace a route to a host since
the command returns the host that discards the packet if it does 
not reach the recipient.

.RS 5
The available \fIoptions\fR are:
.RE

.RS
.TP 20
.BI "-timeout " time
The timeout \fItime\fR is given in seconds with a default of 5 seconds. 
.TP
.BI "-retries " number
Unanswered icmp echo request will be resent \fInumber\fR times with a 
default of 2. 
.TP
.BI "-delay " time
Icmp packets will be send with a minimum delay between successive
packets. The default delay is 0 milliseconds. You can change
this delay using this option to \fItime\fR milliseconds. Delays are
useful to reduce load on intermediate hosts like gateways or bridges.
.TP
.BI "-size " number
Set the size of outgoing icmp packets to \fInumber\fR bytes. The default
packet size is 64 bytes. Sizes smaller than 64 bytes are silently rounded 
to 64 bytes.
.RE

.SS DOMAIN NAME SYSTEM (DNS)
.TP
\fBdns\fR [\fIoptions\fR] \fBaddress\fR \fIhost\fR
The dns command sends a query to the domain name system (DNS) resolver 
for the specified \fIhost\fR. The first form of this command asks for
the ip address contained in the A record. If \fIhost\fR is something 
like ra.ibr.cs.tu-bs.de, you will get the result 134.169.34.12 
134.169.246.34.

.TP 
\fBdns\fR [\fIoptions\fR] \fBptr\fR \fIname\fR
Send a query for the domain name pointer. Pointer records are used
to map an ip address into a fully qualified host name. If \fIname\fR
is something like 134.169.34.12, then the result schould be RA.IBR.CS.TU-BS.DE.

.TP
\fBdns\fR [\fIoptions\fR] \fBhinfo\fR \fIname\fR
Retrieve the host information record. If \fIname\fR is something like 
ra.ibr.cs.tu-bs.de, you will get the result SUN-4/75 SunOS.

.TP
\fBdns\fR [\fIoptions\fR] \fBmx\fR \fIname\fR
Send a query for a mail exchanger record (MX). If \fIname\fR is a
domain name for which an MX record exists, you will get a list of mail
exchanger host. Each element contains a hostname and its priority. For
example, when called with the domain name ibr.cs.tu-bs.de, you will
get {ra.ibr.cs.tu-bs.de 0} {jupiter.ibr.cs.tu-bs.de 10}.

.TP
\fBdns\fR [\fIoptions\fR] \fBsoa\fR \fIname\fR
Ask the DNS for the start of authority (SOA). If \fIname\fR
is something like cs.tu-bs.de, you will get the answer 
infbssys.ips.cs.tu-bs.de.

.RS 5
The available \fIoptions\fR are:
.RE

.RS 
.TP 20
.BI "-server " server
Send the query to the specified \fIserver\fR
instead to the default server(s) of the system.
.TP
.BI "-timeout " time
Set the timeout to \fItime\fR seconds. 
The default timeout is 2 seconds.
.TP
.BI "-retries " number
Set the number of retries to \fInumber\fR. The default number of 
retries is 2. Please note, that the resolver doubles the last timeout 
after every retry.
.RE

.SS NETWORK TIME PROTOCOL (NTP)
.TP
\fBntp\fR [\fIoptions\fR] \fBhost\fR \fIarrayName\fR
This command sends an ntp version 3 mode 6 request to \fIhost\fR and
writes status information into \fIarrayName\fR. The array will contain 
the elements peer.delay, peer.dispersion, peer.offset, peer.precision,
peer.reach, peer.srcadr, peer.stratum, peer.valid, sys.peer,
sys.precision, sys.refid, sys.rootdelay, sys.rootdispersion,
sys.stratum and sys.system.  See RFC 1305 for the meaning of these
variables.

.RS 5
The available \fIoptions\fR are:
.RE

.RS 
.TP 20
.BI "-timeout " time
Set the timeout to \fItime\fR seconds. 
The default timeout is 2 seconds.
.TP
.BI "-retries " number
Set the number of retries to \fInumber\fR. The default number of 
retries is 2.
.RE

.SS USER DATAGRAM PROTOCOL (UDP)
.TP
.B udp open \fR[\fIport\fR]
The udp commands described below provide access to the udp protocol. 
The first one creates a udp datagram socket and associates it with a 
Tcl file handle which will be returned on command completion. The 
optional argument defines the port number to use. The command 
will search for an unused port if no port is given.

.TP
.B udp connect \fIhost\fR \fIport\fR
Open a datagram file handle that is connected to \fIhost\fR
and \fIport\fR. This makes frequent message exchanges faster since 
there is no need to give the host and port arguments to the send 
command once the udp file handle is connected to a target.

.TP
.B udp send \fIfile\fR [\fIhost port\fR] \fImessage\fR
Send a datagram containing \fImessage\fR. The \fIhost\fR and \fIport\fR 
arguments may not be used if the udp file handle is already connected to
a remote host. If the udp file handle is not connected, you must use
these optional arguments to identify the destination of the datagram.

.TP
.B udp receive \fIfile\fR
Receive a datagram from the udp file handle \fIfile\fR. This command 
blocks until a message is received. In most cases, it might be a good
idea to check for pending datagrams using the fileevent command.

.TP
.B udp close \fIfile\fR
Close the udp file handle \fIfile\fR.

.TP
.B udp info \fR[\fIfile\fR]
Calling udp info without any arguments returns a list of all Tcl file
handles that have been created using the udp create command. Info about
the current connection of a udp file handle \fIfile\fR is returned
by using the optional argument. The result is a list containing the 
source ip address, source port, destination ip address and destination 
port.

.SS TRANSMISSION CONTROL PROTOCOL (TCP)
.TP
.B tcp connect \fIhost\fR \fIport\fR
The tcp commands below are used to manage tcp connections. The tcp 
connect command creates a connection to a remote tcp server. The
returned identifier may be used in commands like read, write, gets, 
puts and close. \fIhost\fR gives the name of the host to connect
to. \fIport\fR gives the desired port.

Even if every socket is opened for both reading and writing, seek must
not be invoked between a read and a write. As Tcl file I/O is
implemented on top of the stdio library, communication over a TCP file
handle may require a call to flush whenever you change the direction
(e.g. after a puts and before a gets on the TCP handle). Note,
although many systems can handle this situation automatically, it is
still a nice idea to call flush for those machines who need it.

.TP
.B tcp listen \fR[\fIport\fR]
A listening tcp socket is set up. If \fIport\fR is empty or 0, then
the system will select a port number for you. The returned identifier 
may be used in commands like accept, shutdown and close. 

.TP
.B tcp accept \fIfile\fR
Accepts a connection on an existing server \fIfile\fR (created with the
tcp listen command) and returns an identifier which may be used in 
commands like read, write, gets, puts and close.

.TP
.B tcp shutdown \fIfile\fR \fIhow\fR
Shutdown a connection for reading, writing or both. If \fIhow\fR
is "0", "read" or "receives", then further receives on \fIfile\fR
will be disabled. If \fIhow\fR is "1", "write" or "sends", then 
further sends on \fIfile\fR will be disabled and eof will be send 
on the socket. If \fIhow\fR is "2", "all" or "both", then further 
sends and receives on \fIfile\fR will be disabled.

.TP
.B tcp close \fIfile\fR
Close the tcp file handle \fIfile\fR.

.TP
.B tcp info \fR[\fIfile\fR]
Calling tcp info without any arguments returns a list of all Tcl file
handles that have been created using the tcp listen and tcp connect
command. Info about the current connection of a tcp file handle 
\fIfile\fR is returned by using the optional argument. The result is a
list containing the source ip address, source port, destination ip
address and destination port.

.SS HYPERTEXT TRANSFER PROTOCOL (HTTP)
Scotty provides some basic access to HTTP server. This allows
applications to retrieve documents with a http uniform ressource
locator (URL) (see RFC 1738 for details about URLs). You can get access 
to documents with non http URLs by using a proxy http server. 
.TP
.B http head \fIurl\fR
This command retrieves the header information of the document with the
given URL \fIurl\fR. The header is returned as a list of key-value pairs.

.TP
.B http get \fIurl\fR \fIfileName\fR
Retrieve the document with the URL \fIurl\fR. The body of the document
is written to the file \fIfileName\fR. The command returns the header
as described by the http head command above.

.TP
.B http post \fIurl\fR \fIdocFileName\fR \fIfileName\fR
Post the document given by \fIdocFileName\fR to the URL \fIurl\fR. The
body of the document is written to the file \fIfileName\fR. The
command returns the header as described by the http head command
above.

.TP
.B http put \fIurl\fR \fIfileName\fR
Put the document under the URL \fIurl\fR. The body of the document is
read from the file \fIfileName\fR. The command returns the header as
described by the http head command above.

.TP
.B http delete \fIurl\fR
Delete the document with the URL \fIurl\fR. The command returns status
information.

.TP
.B http proxy \fR[\fIurl\fR]
All subsequent http requests are send to the proxy server given by
\fIurl\fR. Without the \fIurl\fR argument, the currently used proxy
server is returned. Setting the proxy server to an empty string turns
the proxy feature off.

.TP
.B http server \fR[\fIport\fR]
Turn scotty into a simple http server listening on port \fIport\fR.
If called without a \fIport\fR argument, the currently used port number
or an empty string is returned.

.TP
.B http mime \fItype\fR \fIextension\fR
This option can be used to add mime type definitions to the internal
table that maps file name extension to mime types. If called without
an argument, the list of all mime type mappings is returned. A new
mapping is created by supplying a mime \fItype\fR and the file name
\fIextension\fR.

.TP
.B http bind \fIpattern\fR \fImethod\fR \fR[\fIscript\fR]
This option binds a Tcl script to a http request indicated by
\fImethod\fR where the URL matches \fIpattern\fR. Whenever a HTTP
request is received that matches the pattern, \fIscript\fR is evaluated 
and the returned values is send back to the client. Errors during the
evaluation of \fIscript\fR will result in an error response.

Some % sequences will be substituted before the command is evaluated:

.RS
.TP 6
.B %A
the network address of the client
.TP
.B %P
the URL path requested by the client
.TP
.B %S
the search path contained in the URL path
.RE

.SS SUN REMOTE PROCEDURE CALL (SUNRPC)
.TP
.B sunrpc info \fIhost\fR
Return a list of all registered Sun RPC services on
.I host.
The list elements are strings each containing the program number, program
version, protocol name, port number and the name of one RPC service.
The command fails, if 
.I host
is an incorrect name or if the portmapper is not available.

.TP
.B sunrcp probe \fIhost\fR \fIprogram\fR \fIversion\fR \fIprotocol\fR
You can test a Sun RPC service using the sunrpc probe command. This calls
procedure 0 that simply returns. Additional arguments to sunrpc probe
will be ignored allowing one to write simple loops to test all
registered RPC services of a host.

.TP
.B sunrpc stat \fIhost\fR
Make a rstat RPC call to \fIhost\fR and return a list of name type
value triples describing the status of the machine. This command fails
if \fIhost\fR is not reachable or does not support the rstat service.

.TP
.B sunrpc mount \fIhost\fR
Get a list of all NFS file systems mounted from \fIhost\fR.

.TP
.B sunrpc exports \fIhost\fR
Get a list of all NFS file systems exported by \fIhost\fR.

.TP
.B sunrpc ether \fIhost\fR \fBopen
.ns
.TP
.B sunrpc ether \fIhost\fR
.ns
.TP
.B sunrpc ether \fIhost\fR \fBclose
Query the etherd daemon. First you have to set up a connection to
the etherd on \fIhost\fR using sunrpc ether \fIhost\fR open.
Afterwards you can query the etherd which will return a list of name
type value triples describing the packets counted since the last
query. You should close the connection using sunrpc ether \fIhost\fR 
close.

.SS SIMPLE NETWORK MANAGEMENT PROTOCOL (SNMP)
Access to the Simple Network Management Protocol (SNMP) version 1 and
version 2 is provided by two new Tcl commands. The first command
handles queries to the MIB definitions and the second is used to
create SNMP sessions. All interactions with SNMP agents are done using
SNMP session handles. The differences between version 1 and version 2
are mostly hidden by this SNMP implementation (e.g. getbulk requests
are mapped to getnext requests and version 1 traps are signalled as
version 2 traps) once you have configured a valid session handle.

SNMP data types are mapped to a Tcl string representation
automatically. Below is a short description how the basic SNMP
types are mapped to Tcl:

.TP 20
OCTET STRING
A primitive OCTET STRING is represended as a number of hexadecimal
octets separated by colons, e.g. 6e:61:73:65.
.TP
OBJECT IDENTIFIER
An OBJECT IDENTIFIER is represented in the dotted notation. Each 
subidentifier is converted to an unsigned integer. The integer 
subidentifier are separated by dots, e.g. 1.3.6.1.2.1.1. It is
also allowed to have hexadicimal octets as subidentifiers if they
are preceeded by a colon instead of a dot, e.g. 1.3.6.1:02:01:01.
.TP
IpAddress
An IpAddress is converted to the usual dotted notation, e.g. 134.169.34.15.
.TP
TimeTicks
A TimeTicks value is converted to a string of the form nd hh:mm:ss:cc
where n is the number of days, hh is the number hours, mm is the
number of minutes, ss is the number of seconds and cc is the number of
hundreds of a seconds, e.g. 7d 2:04:37.84.
.TP
INTEGER, Integer32
.ns
.TP
Counter, Counter32
.ns
.TP
Gauge, Gauge32
All these numeric values are converted into an integer number. Note,
the label of an enumeration will be returned instead of a number if
there is an enumeration defined for the value of an object.
.TP
Counter64
A Counter64 value is mapped to an integer value on 64 Bit machines
and to a floating point value on 32 Bit machines. 
.PP

Display hints and enumerations defined in textual conventions are also
applied automatically. This allows to use the application specific
representation or the label of an enumeration in your Tcl code.
Explicit conversions are possible using the mib format and mib scan

debian-bo-master/bo/source/net/scotty-2.0.2/scotty.c
/*
 * scotty.c
 *
 * This is scotty, a simple tcl interpreter with some special commands
 * to get information about TCP/IP networks. 
 *
 * Copyright (c) 1993, 1994, 1995
 *
 * J. Schoenwaelder
 * TU Braunschweig, Germany
 * Institute for Operating Systems and Computer Networks
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that this copyright
 * notice appears in all copies.  The University of Braunschweig
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#include "scotty.h"

#ifdef HAVE_TCLX
extern int
TclX_Init		_ANSI_ARGS_((Tcl_Interp *interp));
#endif

static int ined = 0;            /* dont initialize for tkined */
static char *progname;          /* the name of the game */

static Tcl_Interp *interp;      /* the main tcl interpreter */

static Tcl_DString command;	/* Used to assemble lines of terminal input
				 * into Tcl commands. */
static int tty;			/* Non-zero means standard input is a
				 * terminal-like device.  Zero means it's
				 * a file. */

static char *rcFileName = "~/.scottyrc";

/*
 * Forward declarations for procedures defined later in this file:
 */

static void
EventLoop		_ANSI_ARGS_((Tcl_Interp *interp));

/*
 * Taken from the tkMain.c file:
 */

static void
Prompt			_ANSI_ARGS_((Tcl_Interp *interp, int partial));

static void		
StdinProc		_ANSI_ARGS_((ClientData clientData, int mask));

/*
 * This is the main event loop. It does one tk event and checks whether
 * there are command left send from the tkined editor.
 */

static void
EventLoop (interp)
     Tcl_Interp *interp;
{
    while (Tk_DoOneEvent (0)) {
	/* empty loop body */
    }
}

/*
 * Initialize a new Tcl interpreter.
 */

int
Tcl_AppInit (interp)
    Tcl_Interp *interp;
{
    Tcl_SetVar(interp, "scotty_RcFileName", 
	       rcFileName ? rcFileName : "", TCL_GLOBAL_ONLY);

    if (Tcl_Init (interp) != TCL_OK) {
        return TCL_ERROR;
    }

#ifdef HAVE_TCLX
    if (TclX_Init (interp) != TCL_OK) {
        return TCL_ERROR;
    }
#endif

    if (Tk_EventInit (interp) != TCL_OK) {
	return TCL_ERROR;
    }

    if (Scotty_Init (interp) != TCL_OK) {
        return TCL_ERROR;
    }
    
    return TCL_OK;
}

/*
 * here we start
 */

int
main(argc, argv)
    int argc;
    char **argv;
{
    char *fileName = NULL;
    char *query = NULL;
    char buffer[20];
    Tcl_DString sc_argv;
    int i, c;
    int sc_argc = 0;
    int errflg = 0;
    int res;
    int expectOption = 1;

    if ((progname = strrchr(argv[0], '/')) == NULL) {
	progname = argv[0];
    } else {
	progname++;
    }

    Tcl_DStringInit (&sc_argv);

    for (i = 1; i < argc; i++) {
        if (!expectOption || argv[i][0] != '-') {
	    expectOption = 0;
	    Tcl_DStringAppendElement (&sc_argv, argv[i]); 
	    sc_argc++;
	    continue;
	}
	for (c = 1; (i < argc) && (c < strlen(argv[i])); c++) {
	    switch (argv[i][c]) {
	      case 'c' :
		if ( (argv[i][c+1] == '\0') && (++i < argc) ) {
		    query = argv[i];
		    c = strlen(argv[i]);
		} else errflg++;
		break;
	      case 'f':
		if ( (argv[i][c+1] == '\0') && (++i < argc) ) {
		    fileName = argv[i]; 
		    c = strlen(argv[i]);
		} else errflg++;
		break;
	      case 'n':
		rcFileName = NULL;
		break;
	      case 'i':
		ined++;
		break;
	      case '?':
		errflg++;
		break;
	      case '-':
		c = strlen(argv[i]);
		for (i++; i < argc; i++) {
		    Tcl_DStringAppendElement (&sc_argv, argv[i]);
		    sc_argc++;
		}
		break;
	      default:
		c = strlen(argv[i]);
		Tcl_DStringAppendElement (&sc_argv, argv[i]);
		sc_argc++;
		break;
	    }
	}
    }

    if (errflg) {
	fprintf (stderr, "usage: %s [-c query] [-f file] [-n] [-i]\n", 
		 progname);
	return 2;
    }


    /* Create and initialize the Tcl interpreter. */

    interp = Tcl_CreateInterp();
#ifdef TCL_MEM_DEBUG
    Tcl_InitMemory(interp);
#endif

    /* Set the tcl_interactive variable. */

    tty = isatty (0);
    Tcl_SetVar(interp, "tcl_interactive",
	    ((fileName == NULL) && tty) ? "1" : "0", TCL_GLOBAL_ONLY);

    /* Register all new commands. */

    if (Tcl_AppInit(interp) == TCL_ERROR) {
	fprintf(stderr, "Tcl_AppInit failed: %s\n", interp->result);
        return TCL_ERROR;
    }

    /* Set the global variables argc and argv. */

    Tcl_SetVar (interp, "argv", Tcl_DStringValue(&sc_argv), TCL_GLOBAL_ONLY);
    Tcl_SetVar (interp, "argv0", (fileName != NULL) ? fileName : argv[0],
		TCL_GLOBAL_ONLY);
    Tcl_DStringFree (&sc_argv);
    sprintf (buffer, "%d", sc_argc);
    Tcl_SetVar (interp, "argc", buffer, TCL_GLOBAL_ONLY);

    /*
     * Process a single command.
     */

    if (query) {
	int result = Tcl_Eval (interp, query);
	if (result == TCL_OK) {
	    if (*interp->result != 0) {
		printf ("%s\n", interp->result);
	    }
	} else {
	    fprintf (stderr, "Error: %s\n%s\n", interp->result,
		     Tcl_GetVar (interp, "errorInfo", TCL_GLOBAL_ONLY));
	}
	EventLoop (interp);
	return 0;
    }
    
    /*
     * Process queries from a script file.
     */

    if (fileName) {

	if (ined) {
	    if (Tcl_Eval (interp, "ined size") != TCL_OK) {
		fprintf (stderr, "Fatal: can not talk to tkined: %s\n",
			 interp->result);
		exit (32);
	    }
	}

	res = Tcl_EvalFile (interp, fileName);
	if (res != TCL_OK) {
	    fprintf (stderr, "Error: %s\n", interp->result);
	}

	EventLoop (interp);
	return res;
    }
    
    /*
     * We are interactive -- that's where the fun beginns.
     */

    if (tty) {
	Tcl_SetVar (interp, "tcl_interactive", "1", TCL_GLOBAL_ONLY);
    }
    fflush (stdout);

    Tcl_DStringInit(&command);
    Tk_CreateFileHandler(0, TK_READABLE, StdinProc, (ClientData) 0);
    if (tty) {
	Prompt(interp, 0);
    }

    EventLoop (interp);
    return 0;
}


/*
 *----------------------------------------------------------------------
 *
 * StdinProc --
 *
 *	This procedure is invoked by the event dispatcher whenever
 *	standard input becomes readable.  It grabs the next line of
 *	input characters, adds them to a command being assembled, and
 *	executes the command if it's complete.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Could be almost arbitrary, depending on the command that's
 *	typed.
 *
 *----------------------------------------------------------------------
 */

    /* ARGSUSED */
static void
StdinProc(clientData, mask)
    ClientData clientData;		/* Not used. */
    int mask;				/* Not used. */
{
#define BUFFER_SIZE 4000
    char input[BUFFER_SIZE+1];
    static int gotPartial = 0;
    char *cmd;
    int code, count;

    count = read(fileno(stdin), input, BUFFER_SIZE);
    if (count <= 0) {
	if (!gotPartial) {
	    if (tty) {
		Tcl_Eval(interp, "exit");
		exit(1);
	    } else {
		Tk_DeleteFileHandler(0);
	    }
	    return;
	} else {
	    count = 0;
	}
    }
    cmd = Tcl_DStringAppend(&command, input, count);
    if (count != 0) {
	if ((input[count-1] != '\n') && (input[count-1] != ';')) {
	    gotPartial = 1;
	    goto prompt;
	}
	if (!Tcl_CommandComplete(cmd)) {
	    gotPartial = 1;
	    goto prompt;
	}
    }
    gotPartial = 0;

    /*
     * Disable the stdin file handler while evaluating the command;
     * otherwise if the command re-enters the event loop we might
     * process commands from stdin before the current command is
     * finished.  Among other things, this will trash the text of the
     * command being evaluated.
     */

    Tk_CreateFileHandler(0, 0, StdinProc, (ClientData) 0);
    code = Tcl_RecordAndEval(interp, cmd, 0);
    Tk_CreateFileHandler(0, TK_READABLE, StdinProc, (ClientData) 0);
    Tcl_DStringFree(&command);
    if (*interp->result != 0) {
	if ((code != TCL_OK) || (tty)) {
	    /*
	     * The statement below used to call "printf", but that resulted
	     * in core dumps under Solaris 2.3 if the result was very long.
	     */

	    puts(interp->result);
	}
    }

    /*
     * Output a prompt.
     */

    prompt:
    if (tty) {
	Prompt(interp, gotPartial);
    }
    Tcl_ResetResult(interp);
}

/*
 *----------------------------------------------------------------------
 *
 * Prompt --
 *
 *	Issue a prompt on standard output, or invoke a script
 *	to issue the prompt.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	A prompt gets output, and a Tcl script may be evaluated
 *	in interp.
 *
 *----------------------------------------------------------------------
 */

static void
Prompt(interp, partial)
    Tcl_Interp *interp;			/* Interpreter to use for prompting. */
    int partial;			/* Non-zero means there already
					 * exists a partial command, so use
					 * the secondary prompt. */
{
    char *promptCmd;
    int code;

    promptCmd = Tcl_GetVar(interp,
	partial ? "tcl_prompt2" : "tcl_prompt1", TCL_GLOBAL_ONLY);
    if (promptCmd == NULL) {
	defaultPrompt:
	if (!partial) {
	    fputs("% ", stdout);
	}
    } else {
	code = Tcl_Eval(interp, promptCmd);
	if (code != TCL_OK) {
	    Tcl_AddErrorInfo(interp,
		    "\n    (script that generates prompt)");
	    fprintf(stderr, "%s\n", interp->result);
	    goto defaultPrompt;
	}
    }
    fflush(stdout);
}
